overlord/snapstate/snapstate.go

```diff
@@ -1358,6 +1358,117 @@ func installWithDeviceContext(st *state.State, name string, opts *RevisionOption
+		sar, err := downloadInfo(ctx, st, name, ro, userID, dc)
+		if err != nil {
+			return nil, "", err
+		}
+		return sar.Info, sar.RedirectChannel, nil
+	}
+
+	return download(st, name, opts, userID, flags, deviceCtx, snapDownloadInfo)
+}
+
+func download(st *state.State, name string, opts *RevisionOptions, userID int, flags Flags, deviceCtx DeviceContext, snapDownloadInfo snapInfoForDownload) (*state.TaskSet, error) {
+	if opts == nil {
+		opts = &RevisionOptions{}
+	}
+
+	if opts.CohortKey != "" && !opts.Revision.Unset() {
+		return nil, errors.New("cannot specify revision and cohort")
+	}
+
+	if opts.Channel == "" {
+		opts.Channel = "stable"
+	}
+
+	var snapst SnapState
+	err := Get(st, name, &snapst)
+	if err != nil && !errors.Is(err, state.ErrNoState) {
+		return nil, err
+	}
+
+	// if the snap is already installed, then it should already be downloaded
+	// somewhere
+	if snapst.IsInstalled() {
+		return nil, &snap.AlreadyInstalledError{Snap: name}
+	}
+
+	if err := snap.ValidateInstanceName(name); err != nil {
+		return nil, fmt.Errorf("invalid instance name: %v", err)
+	}
+
+	info, redirectChannel, err := snapDownloadInfo(deviceCtx, opts)
+	if err != nil {
+		return nil, err
+	}
+
+	if flags.RequireTypeBase && info.Type() != snap.TypeBase && info.Type() != snap.TypeOS {
+		return nil, fmt.Errorf("unexpected snap type %q, instead of 'base'", info.Type())
+	}
+
+	// TODO: should changeKind be "download"? i don't see that as one of the
+	// options that this function operates on
+	if err := checkDiskSpace(st, "install", []minimalInstallInfo{installSnapInfo{info}}, userID); err != nil {
+		return nil, err
+	}
+
+	providerContentAttrs := defaultProviderContentAttrs(st, info)
+	snapsup := &SnapSetup{
+		Channel:            opts.Channel,
+		Base:               info.Base,
+		Prereq:             getKeys(providerContentAttrs),
+		PrereqContentAttrs: providerContentAttrs,
+		UserID:             userID,
+		Flags:              flags.ForSnapSetup(),
+		DownloadInfo:       &info.DownloadInfo,
+		SideInfo:           &info.SideInfo,
+		Type:               info.Type(),
+		Version:            info.Version,
+		PlugsOnly:          len(info.Slots) == 0,
+		InstanceKey:        info.InstanceKey,
+		auxStoreInfo: auxStoreInfo{
+			Media: info.Media,
+			// TODO: still needed here?
+			// XXX we store this for the benefit of old snapd
+			Website: info.Website(),
+		},
+		CohortKey:          opts.CohortKey,
+		ExpectedProvenance: info.SnapProvenance,
+	}
+
+	if redirectChannel != "" {
+		snapsup.Channel = redirectChannel
+	}
+
+	return doDownload(st, snapsup)
+}
+
+func doDownload(st *state.State, snapsup *SnapSetup) (*state.TaskSet, error) {
+	revisionStr := fmt.Sprintf(" (%s)", snapsup.Revision())
+
+	prereq := st.NewTask("prerequisites", fmt.Sprintf(i18n.G("Ensure prerequisites for %q are available"), snapsup.InstanceName()))
+	prereq.Set("snap-setup", snapsup)
```

Comment:
I think for our use cases we we can't and don't want to generate a prerequisites task as it might create actual install tasks for those. As recovery system need to be self contained we whould not need to go find prereqs anyway, otoh we will have to check that they are all already accounted for

Reply:
Do you think that checking for the existence of prerequisites should happen in `Download` or should it happen from the caller's code?

Reply:
the caller, I'm not sure how would you check from Download itself. The prerequisites checking needing is similar, maybe slightly simpler that the one in remodel case

Reply:
Okay, I've removed installing the prerequisites from this function.